app/app.py

```diff
@@ -18,11 +19,19 @@ def index():
 
 @app.route('/ask', methods=['POST'])
 def ask():
+    # Calculate the sha256 of the concatenated string of 32-bit X-Nonce header and raw body.
+    x_nonce = request.headers.get('X-Nonce')
+    x_nonce_bytes = int(x_nonce).to_bytes(4, byteorder='little')  # 32-bit = 4 bytes
+    calculated_hash = hashlib.sha256(x_nonce_bytes + request.get_data()).digest()
+    first_uint32 = int.from_bytes(calculated_hash[:4], byteorder='big')
+    if first_uint32 > 0x50000:
+        return jsonify({"error": "Invalid proof of work"}), 400
+
```

Comment:
I would prefer we do this in another function and expose a parameter that lets us modulate the limit.

```
def validate_pow(data, nonce, difficulty):
    # Calculate the sha256 of the concatenated string of 32-bit X-Nonce header and raw body.
    nonce_bytes = int(nonce).to_bytes(4, byteorder='little')  # 32-bit = 4 bytes
    calculated_hash = hashlib.sha256(nonce_bytes + request.get_data()).digest()
    first_uint32 = int.from_bytes(calculated_hash[:4], byteorder='big')
    return first_uint32 <= 0x50000
```

Then we can

```suggestion
    if !validate_pow(request.headers.get('X-Nonce')
, request.get_data(), 0x50000)
        return jsonify({"error": "Invalid proof of work"}), 400

```

Reply:
done.